### PR TITLE
MovieLens Datasets

### PR DESCRIPTION
**Pull request by Dsantra92**
- [x] MovieLens Struct
- [x] Base.show
- [x] Docs
- [x] MovieLens 100K
- [x] MovieLens 1m
- [x] MoviLens 20m
- [x] MovieLens 25m
- [x] Tests

### CODE REVIEW

File: src/datasets/graphs/movielens.jl [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+  user_df = read_csv_asdf(joinpath(dir, user_data_file), header=false)
+
+  user_ids = user_df[!, 1]
+  @assert minimum(user_ids) == 1
+  user_data["num_users"] = maximum(user_ids)
+  user_data["age"] = user_df[!, 2]
+  user_data["gender"] = user_df[!, 3] .== "M" # I hope I don't get cancelled for binarizing this field
+  user_data["occupation"] = user_df[!, 4]
+  user_data["zipcode"] = user_df[!, 5]
+  return user_data
```
Comment (CarloLucibello) [inline src/datasets/graphs/movielens.jl:57]: The indentation is not uniform in this file. It should be 4 blanks everywhere

Comment (Dsantra92) [reply src/datasets/graphs/movielens.jl:57]: Yeah, vim messed it up somewhere, I will fix these things in a cleanup commit later.

**Comment by Dsantra92** [comment]
Forgot to add the indentation fix change, will be added in later commit.

**Comment by Dsantra92** [comment]
There are inconsistencies in data storing format across the 3 variations: 100k,1m and current datasets(20m, 25m etc.). Will address the issue when all of them have working APIs.

File: src/datasets/graphs/movielens.jl [hunk 2]
```diff
@@ -0,0 +1,538 @@
...
+
+There are 6 versions of movielens datasets currently supported: "25m", "20m", "latest-small", "10m", "1m", "100k". 
+The 1m and 100k datasets contains movies data and ratings data along with demographic data.
+Starting from 10m datset, Movielens datasets no longer contain the demographic data. 
+These datassets contain movie data, rating data and tag inoformation. 
+
+The 20m, 25m datasets additonally contain [genome tag scores](http://files.grouplens.org/papers/tag_genome.pdf). 
+Each movie in theses datasets contains tag relevance scores for every tag.
+
+`movie` and `user` represent the movie and user nodes in the graph. The rating is represented by an edge between them : `(user, rating, movie)`.
```
Comment (CarloLucibello) [inline src/datasets/graphs/movielens.jl:29]:
```suggestion
Each dataset contains an heterogeneous graph, with two kinds of nodes, 
`movie` and `user`. The rating is represented by an edge between them: `(user, rating, movie)`. 
```

File: src/datasets/graphs/movielens.jl [hunk 3]
```diff
@@ -0,0 +1,538 @@
...
+The 1m and 100k datasets contains movies data and ratings data along with demographic data.
+Starting from 10m datset, Movielens datasets no longer contain the demographic data. 
+These datassets contain movie data, rating data and tag inoformation. 
+
+The 20m, 25m datasets additonally contain [genome tag scores](http://files.grouplens.org/papers/tag_genome.pdf). 
+Each movie in theses datasets contains tag relevance scores for every tag.
+
+`movie` and `user` represent the movie and user nodes in the graph. The rating is represented by an edge between them : `(user, rating, movie)`. 
+20m, 25m and latest-small datasets also contain `tag` nodes. The relation between the other nodes are represented using edges: `(user, tag, movie)` and 
+optionally `(movie, score, tag)`.
```
Comment (CarloLucibello) [inline src/datasets/graphs/movielens.jl:31]:
```suggestion
20m, 25m, and latest-small datasets also contain `tag` nodes and edges of type `(user, tag, movie)` and 
optionally `(movie, score, tag)`.
```

File: src/datasets/graphs/movielens.jl [hunk 4]
```diff
@@ -0,0 +1,538 @@
...
+ ⋮
+```
+
+## References
+
+[1] [GroupLens Website](https://grouplens.org/datasets/movielens/)
+[2] [TensorFlow MovieLens Implementation](https://www.tensorflow.org/datasets/catalog/movielens)
+[3] Jesse Vig, Shilad Sen, and John Riedl. 2012. The Tag Genome: Encoding Community Knowledge to Support Novel Interaction. ACM Trans. Interact. Intell. Syst. 2, 3, Article 13 (September 2012), 44 pages. https://doi.org/10.1145/2362394.2362395
+
+[4] F. Maxwell Harper and Joseph A. Konstan. 2015. The MovieLens Datasets: History and Context. ACM Trans. Interact. Intell. Syst. 5, 4, Article 19 (January 2016), 19 pages. https://doi.org/10.1145/2827872
```
Comment (CarloLucibello) [inline src/datasets/graphs/movielens.jl:120]:
```suggestion
[1] [GroupLens Website](https://grouplens.org/datasets/movielens/)   
[2] [TensorFlow MovieLens Implementation](https://www.tensorflow.org/datasets/catalog/movielens)   
[3] Jesse Vig, Shilad Sen, and John Riedl. 2012. The Tag Genome: Encoding Community Knowledge to Support Novel Interaction. ACM Trans. Interact. Intell. Syst. 2, 3, Article 13 (September 2012), 44 pages. https://doi.org/10.1145/2362394.2362395.   
[4] F. Maxwell Harper and Joseph A. Konstan. 2015. The MovieLens Datasets: History and Context. ACM Trans. Interact. Intell. Syst. 5, 4, Article 19 (January 2016), 19 pages. https://doi.org/10.1145/2827872  
```

File: src/datasets/graphs/movielens.jl [hunk 5]
```diff
@@ -0,0 +1,538 @@
...
+        metadata = get_movielens_metadata(data)
+        return MovieLens(name, metadata, [g])
+    elseif name in ["20m", "25m", "latest-small"]
+        data = read_current_data(dir)
+        g = generate_movielens_graph(data...)
+        metadata = get_movielens_metadata(data)
+        return MovieLens(name, metadata, [g])
+    else
+        error("Functionality for ml-$name has not been implemented yet")
+    end
```
Comment (CarloLucibello) [inline src/datasets/graphs/movielens.jl:155]:
```suggestion
    if name == "100k"
        data = read_100k_data(dir)
    elseif name == "1m"
        data = read_1m_data(dir)
    elseif name == "10m"
        data = read_10m_data(dir)
    elseif name in ["20m", "25m", "latest-small"]
        data = read_current_data(dir)
    else
        error("Functionality for ml-$name has not been implemented yet")
    end
    g = generate_movielens_graph(data...)
    metadata = get_movielens_metadata(data)
    return MovieLens(name, metadata, [g])
```

**Pull request merged**